scripts/scheduler_ingest.py

```diff
@@ -0,0 +1,159 @@
+        """
+        Print each 'start' attribute within a list, prefixed by an index.
+        :param data: The list of items to print the start dates of
+        """
+        for idx, item in enumerate(data):
+            print(f"{idx}: {item['start']}")
+
+    def _pre_process(self, raw_cycle_data, raw_maintenance_data):
+        """
+        Makes adjustments to the data without converting it to a different data type.
+        These adjustments are necessary for the data to be properly processed
+        into Cycle and Maintenance Day objects.
+        :param raw_cycle_data: Cycle data received from the Scheduler API (as a list)
+        :param raw_maintenance_data: Maintenance day data received from the Scheduler API (as a list)
+        :return: A single list containing all valid items from both lists
+        """
+        data = self._combine_lists(raw_cycle_data, raw_maintenance_data)
+        data = self._sort_by_date(data)
+        data = self._clean_data(data)
+        return data
+
+    def _clean_data(self, data):
+        """
+        Removes invalid items from the data.
+        :param data: The list of items to be validated
+        :return: The list of items that were passed in, but with all invalid items removed
+        """
+        print("cleaning..")
+        print(f"Original length: {len(data)}")
+        clean_list = []
+        for item in data:
+            if 'name' in item and not re.search(self._cycle_name_regex, item['name']):
+                print(f"\nItem removed due to strange cycle name ({item['name']}).\nFull item: {item}")
+            elif item['start'].replace(tzinfo=None) < self._earliest_possible_date:
+                print(f"\nItem removed due to impossible date ({item['start']}).\nFull item: {item}")
+            else:
+                clean_list.append(item)
+        print(f"New length: {len(clean_list)}")
+        return clean_list
+
+    # TODO: question - I've used the method of combining the lists as this was suggested,
+    #  but I'm not exactly sure why we do this instead of keeping them separate?
+    #  Is it a speed optimisation perhaps? I wonder if it makes us more prone to bugs/confusing code..
+    @staticmethod
+    def _combine_lists(first, second):
+        """
+        Combines two lists together.
+        :param first: The first list to be combined
+        :param second: The second list to be combined
+        :return: The single combined list
+        """
+        print("combining..")
+        combined = first + second
+        return combined
+
+    def _sort_by_date(self, data):
+        """
+        Sorts a list by a date field (specified by the _sort_by_field variable)
+        :param data: The list to be sorted
+        :return: The list sorted by the specified field
+        """
+        print("sorting..")
+        return sorted(data, key=lambda date: date[self._sort_by_field])
+
+    # TODO: question - can a maintenance day ever occur outside of a cycle period?
```

Comment:
They shouldn't and I think we should trust that this is validated at the business apps end. Although could you add a comment to say that we are making this assumption?